src/types.ts

```diff
@@ -1,32 +1,36 @@
-import type { ExtractFields, QueryIO } from './generated/types'
+import type { EntityQueryMap } from './generated/types'
 
 type QueryInput = {
-  [K in keyof QueryIO]?: QueryIO[K]['input'];
+  [K in keyof EntityQueryMap]?: EntityQueryMap[K]['input'];
 }
 
+type ExtractFields<T, F extends (keyof T)[] | undefined> = F extends (keyof T)[]
+  ? Pick<T, F[number]>
+  : T
+
 // These types are for internal use and shouldn't be exported in index.ts
 type QueryOutput<T extends QueryInput> = {
-  [K in keyof T & keyof QueryIO]: T[K] extends { fields: (keyof QueryIO[K]['entity'])[] }
-    ? QueryIO[K]['wrapper'] extends 'array'
-      ? ExtractFields<QueryIO[K]['entity'], T[K]['fields']>[]
-      : ExtractFields<QueryIO[K]['entity'], T[K]['fields']>
-    : QueryIO[K]['output'];
+  [K in keyof T & keyof EntityQueryMap]: T[K] extends { fields: (keyof EntityQueryMap[K]['entity'])[] }
+    ? EntityQueryMap[K]['wrapper'] extends 'array'
+      ? ExtractFields<EntityQueryMap[K]['entity'], T[K]['fields']>[]
+      : ExtractFields<EntityQueryMap[K]['entity'], T[K]['fields']>
+    : EntityQueryMap[K]['output'];
```

Comment:
is there any way to make this more readable? Maybe not, just asking because it's very hard to read this peace of code